cranelift/wasm/src/code_translator.rs

```diff
@@ -3579,50 +3581,260 @@ where
+
+    // If the number of bytes to be tagged is larger than this threshhold,
+    // insert a loop that tags using `st2g` chunks.
+    // Threshhold should be a multiple of 32, so each `st2g` chunk can be
+    // exactly divided into only 32-byte tagging `st2g` instructions
+    // This constant was adapted from clang.
+    const LOOP_UNROLL_BYTES_THRESHHOLD_I64: i64 = 160;
+    const LOOP_UNROLL_BYTES_THRESHHOLD_U64: u64 = 160;
+
     // Assert that `size` is 16 byte aligned, and trap if not
     if size % 16 != 0 {
         return Err(WasmError::User(format!(
             "size must be 16 byte aligned, was {size}"
         )));
     }
 
-    // First insert an overflow check. This allows us to insert non-trapping instructions later.
-    let c = builder.ins().iconst(I64, size);
+    // TODO: maybe optimize by adding `size` statically as immediate
+    // Trap if (iter_ptr + size) overflows, so we can use non-trapping instructions later
     builder
-    .ins()
-    .uadd_overflow_trap(iter_ptr, c, ir::TrapCode::HeapOutOfBounds);
+        .ins()
+        .uadd_overflow_trap(iter_ptr, size_dynamic, ir::TrapCode::HeapOutOfBounds);
 
-    let mut i = 0;
-    while size - i >= 32 {
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_st2g(tagged_ptr, iter_ptr);
-        i += 32;
-    }
-    if i < size {
-        let iter_ptr = builder.ins().iadd_imm(iter_ptr, i);
-        builder.ins().arm64_stg(tagged_ptr, iter_ptr);
+    let size_u64: u64 = size as u64;
+    // In Rust, for u64: (a / b) == floor(a / b), which we make use of here
+    let num_required_st2g_loops: u64 = size_u64 / LOOP_UNROLL_BYTES_THRESHHOLD_U64;
+    let num_required_last_stgs: u64 = (size_u64 % LOOP_UNROLL_BYTES_THRESHHOLD_U64) / 16;
+
+    // TODO: cast may be wrong, but i64 is required for use as a Imm64
+    let num_required_st2g_loops_i64: i64 = num_required_st2g_loops as i64;
+
+    if num_required_st2g_loops <= 1 {
+        // WITHOUT_LOOP_UNROLLING_THRESHHOLD
+
+        tag_memory_region_static_without_loop_unrolling_threshhold(
+            iter_ptr, 0, tagged_ptr, size, builder,
+        );
+    } else {
+        // WITH_LOOP_UNROLLING_THRESHHOLD
+
+        // === Block definitions
+        // st2g_loop_condition_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_condition_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: does this really require iter_ptr and counter as parameters, since they are never modified in the condition
+        // st2g_loop_body_block(iter_ptr, st2g_loop_counter)
+        let st2g_loop_body_block =
+            block_with_params(builder, [ValType::I64, ValType::I64], environ)?;
+
+        // TODO: potential optimization: since this isn't required for all generated code, can we conditionally define this block only when needed?
+        // last_stgs_block()
+        let last_stgs_block = block_with_params(builder, [], environ)?;
```

Comment:
I could see wasmtime/cranelift already doing this (i.e. removing a basic block that I declared but never used), but I'm not sure.

Reply:
Yes, I think it's better if we only generate this block if we need it.

Reply:
Should only be generated when required now.